onnxruntime/core/providers/cuda/math/softmax_impl.cu

```diff
@@ -39,67 +39,59 @@ Status dispatch_warpwise_softmax_forward(cudaStream_t stream, output_t* dst, con
-        break;
-      case 6:  // 64
-        softmax_warp_forward<input_t, output_t, acc_t, 6, is_log_softmax>
-            <<<blocks, threads, 0, stream>>>(dst, src, batch_count, softmax_elements_stride, softmax_elements);
-        break;
-      case 7:  // 128
-        softmax_warp_forward<input_t, output_t, acc_t, 7, is_log_softmax>
-            <<<blocks, threads, 0, stream>>>(dst, src, batch_count, softmax_elements_stride, softmax_elements);
-        break;
-      case 8:  // 256
-        softmax_warp_forward<input_t, output_t, acc_t, 8, is_log_softmax>
-            <<<blocks, threads, 0, stream>>>(dst, src, batch_count, softmax_elements_stride, softmax_elements);
-        break;
-      case 9:  // 512
-        softmax_warp_forward<input_t, output_t, acc_t, 9, is_log_softmax>
-            <<<blocks, threads, 0, stream>>>(dst, src, batch_count, softmax_elements_stride, softmax_elements);
-        break;
-      case 10:  // 1024
-        softmax_warp_forward<input_t, output_t, acc_t, 10, is_log_softmax>
-            <<<blocks, threads, 0, stream>>>(dst, src, batch_count, softmax_elements_stride, softmax_elements);
-        break;
-      default:
-        break;
-    }
-  }
+
+#define LAUNCH_KERNEL(kernel_name, log2_elements_value)                                                                  \
+  kernel_name<input_t, output_t, acc_t, log2_elements_value, is_log_softmax>                                             \
+    <<<blocks, threads, shared_memory_size, stream>>>(dst, src, batch_count, softmax_elements_stride, softmax_elements);
+
+#define CASE_LOG2_ELEMENTS(log2_elements_value)                                                                          \
+  case log2_elements_value: {                                                                                            \
+    if constexpr (log2_elements_value <= 10) {                                                                                     \
+      LAUNCH_KERNEL(softmax_warp_forward, log2_elements_value)                                                           \
+    } else {                                                                                                             \
+      LAUNCH_KERNEL(softmax_warp_forward_resource_efficient, log2_elements_value)                                        \
+    }                                                                                                                    \
+  } break
+
+        CASE_LOG2_ELEMENTS(0);
+        CASE_LOG2_ELEMENTS(1);
+        CASE_LOG2_ELEMENTS(2);
+        CASE_LOG2_ELEMENTS(3);
+        CASE_LOG2_ELEMENTS(4);
+        CASE_LOG2_ELEMENTS(5);
+        CASE_LOG2_ELEMENTS(6);
+        CASE_LOG2_ELEMENTS(7);
+        CASE_LOG2_ELEMENTS(8);
+        CASE_LOG2_ELEMENTS(9);
+        CASE_LOG2_ELEMENTS(10);
+        CASE_LOG2_ELEMENTS(11); // start to use softmax_warp_forward_resource_efficient instead of softmax_warp_forward for better performance
```

Comment:
For this "// start to use softmax_warp_forward_resource_efficient instead of softmax_warp_forward for better performance".

Before your change, is softmax_warp_forward runnable with more than > 1024 dimension to reduce? I was thinking softmax_warp_forward_resource_efficient saves register usage, then it make running > 1024 dimension reduction feasible within a warp, isn;t it?

Reply:
original version can run with >1024 case, while perf is worse than the new version